### ISSUE TITLE
NetBeans Connector in the Chrome Web Store is outdated

### ISSUE THREAD

**Issue by vieiro** [apache/netbeans#6601]
### Apache NetBeans version

Apache NetBeans latest daily build

### What happened

The [NetBeans Connector in the Chrome Web Store](https://chrome.google.com/webstore/detail/netbeans-connector/hafdlehgocfcodbgjnpecfajgkeejnaa) needs to be updated to work on recent versions of Chrome, if this is possible, or be removed from the Chrome Web Store.

Sources for the connector seem to be still available at [extbrowser/plugins](https://github.com/apache/netbeans/tree/4086bd3bd687c616694d74f21aeab2ff6addc329/webcommon/extbrowser.chrome/plugins).

The message in the IDE log looks like this:

```
INFO [org.netbeans.ui.metrics.web.clientproject]: USG_PROJECT_HTML5_CONFIGURE
INFO [org.netbeans.modules.extbrowser.plugins.chrome.ChromeManagerAccessor.ChromeExtensionManager]: Could not find chrome extension in installation directory!
INFO [org.netbeans.modules.web.webkit.debugging.api.Debugger]: Enable failed: {"id":0,"error":"{\"message\":\"Cannot attach to this target.\"}"}
INFO [org.netbeans.modules.extbrowser.plugins.chrome.ChromeManagerAccessor.ChromeExtensionManager]: Could not find chrome extension in installation directory!
INFO [org.netbeans.modules.web.webkit.debugging.api.Debugger]: Enable failed: {"id":1,"error":"{\"message\":\"Cannot attach to this target.\"}"}
```

### How to reproduce

1. Install the "NetBeans Connector" from the Chrome Web Store.
2. Create an HTML5 project (I've used a HTML5 template)
3. Try to run the "index.html" file.

### Did this work correctly in an earlier version?

No / Don't know

### Operating System

Any operating system (I'm on Linux). This used to work in NB8.2.

### JDK

JDK11, not important.

### Apache NetBeans packaging

Own source build

### Anything else

Issue https://github.com/apache/netbeans/issues/4987 is probably the same problem. Closing it in favor of this one.

### Are you willing to submit a pull request?

No

**Comment by vieiro** [comment]
It seems the plugin has a manifest version 2, but [Chrome cannot run manifest version2 plugins since January, 2023](https://developer.chrome.com/blog/mv2-transition/).

So the plugin should be upgraded to manifest version 3. This will require [replacing the background page with a service worker](https://css-tricks.com/how-to-transition-to-manifest-v3-for-chrome-extensions/) and changing security policy permissions, etc. [This is a nice description](https://css-tricks.com/how-to-transition-to-manifest-v3-for-chrome-extensions/) on the changes required for the upgrade.

Anyone with experience in Chrome plugins wants to tackle the problem?

**Comment by mhalachev** [comment]
Indeed, it would be advisable to upgrade the extension to Manifest V3, preferably along with the next NB release, as Manifest V2 extensions are likely to be phased-out in the not-so-distant future.

https://developer.chrome.com/blog/resuming-the-transition-to-mv3

TL;DR 
> We will begin disabling Manifest V2 extensions in [pre-stable versions of Chrome](https://developer.chrome.com/docs/web-platform/chrome-release-channels) (Dev, Canary, and Beta) as early as June 2024 […] Users [...] will see Manifest V2 extensions automatically disabled in their browser and will no longer be able to install Manifest V2 extensions from the Chrome Web Store. […] We expect it will take at least a month […] before expanding the rollout to stable channel Chrome.

The process will require rewiring some internals while replacing the background page with a service worker.

@mbien Would it be appropriate to get in touch with Jan Stola regarding the extension refactoring?

@viero To date, the extension still functions, with some workarounds. First, the extension shall be manually enabled. Then, When Run Project is selected, NetBeans presents a message “Browser refused to debug this tab…”. However, if “Debug in NetBeans” is clicked from the extension’s menu right away, it performs as expected.

**Comment by mbien** [comment]
> Would it be appropriate to get in touch with Jan Stola regarding the extension refactoring?

@mhalachev If you want to ask to please refactor the plugin then probably not since too much time flew down the river already ;)

@neilcsmith-net @ebarboni do we have access to the account which manages the plugin at the chrome store? If not: we should at some point try to take the plugin down since its not ideal when it points back to NB and we can't upload a new release.

**Comment by mhalachev** [comment]
@mbien Better late than never ;) The extension is very useful for web frontends and it would be nice to keep.

I attempted some tinkering with this and was able to get a working [Manifest V3](https://developer.chrome.com/docs/extensions/develop/migrate/what-is-mv3) extension to some extent (debugging starts without an error, Browser DOM appears in NetBeans, changes propagate in Chrome, etc.) but the extension UI within the browser is now broken. 

Here is what what the Manifest V3 looks like:
```json
{
    "name": "__MSG__pluginName__",
    "description": "__MSG__pluginDescription__",
    "manifest_version": 3,
    "minimum_chrome_version": "88",
    "version": "1.1.6",
    "homepage_url": "https://chrome.google.com/webstore/detail/netbeans-connector/hafdlehgocfcodbgjnpecfajgkeejnaa",
    "background": {
        "service_worker": "js/background.js"
    },
    "devtools_page": "html/devtools.html",
    "options_page": "html/options.html",
    "action": {
        "default_icon": "img/netbeans16.png",
        "default_title": "Open NetBeans actions",
        "default_popup": "html/popup.html"
    },
    "permissions": [
        "contextMenus",
        "tabs",
        "debugger",
        "storage"
    ],
    "content_security_policy": {
        "extension_pages": "default-src 'self'; connect-src ws://127.0.0.1:8008/"
    },
    "icons": {
        "16": "img/netbeans16.png",
        "48": "img/netbeans48.png",
        "128": "img/netbeans128.png"
    },
    "default_locale": "en"
}
```

The service worker itself:
```javascript
importScripts('../i18n/i18n.js', './common.js', './chrome.js');

chrome.runtime.onStartup.addListener(function() {
    chrome.action.disable();  
});
```
Currently this is only a "duct tape fix", just to get things up and running at all. Ideally, all [extension lifecycle](https://developer.chrome.com/docs/extensions/develop/concepts/service-workers/lifecycle) stuff from common.js and chrome.js should go here. Ref. [Migrate to a service worker](https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers).

Also [replaced some unsupported APIs](https://developer.chrome.com/docs/extensions/develop/migrate/api-calls) and was able at least get the context menu item partially working:

chrome.js, l.91 `chrome.pageAction.show(tabId);` > `chrome.action.enable(tabId);`

...the context menu itself at line 119:

```javascript
    chrome.contextMenus.removeAll(function() {
        chrome.contextMenus.create({
            id: 'selectionMode',
            title: NetBeans.contextMenuName(),
            contexts: ['all'],
            documentUrlPatterns: [NetBeans.contextMenuUrl],
//  onclick does not work anymore with Manifest V3           
//            onclick: function() {
//                NetBeans.setSelectionMode(!NetBeans.getSelectionMode());
//            }
        },
        function() {
            NetBeans.contextMenuCreationInProgress = false;
        });
// Instead, addListener should be used...
        chrome.contextMenus.onClicked.addListener((info, tab) => {
            if (info.menuItemId === 'selectionMode') {
                NetBeans.setSelectionMode(!NetBeans.getSelectionMode());
            }
        })
```

common.js l.182 `chrome.extension.onMessage.addListener` > `chrome.runtime.onMessage.addListener`

This is the progress thus far. While it's not extensive, it might serve as a starting point. :) Although very well crafted, the extension is not overly simplistic, and while seeing it for the first time still need to figure out how exactly things work. 

Hint: The extension can be easily [loaded unpacked](https://developer.chrome.com/docs/extensions/get-started/tutorial/hello-world#load-unpacked) directly from the NB codebase for tinkering.

**Comment by mhalachev** [comment]
A quick update: Chrome now displays an alert for extensions using Manifest V2, indicating that support for these extensions may  soon be discontinued.

![chrome-extensions-alert](https://github.com/user-attachments/assets/db1dea80-4768-4436-9e48-e8bbac1ae1f0)

PS: Find alternative points here: https://chromewebstore.google.com/detail/hafdlehgocfcodbgjnpecfajgkeejnaa/related-recommendations

**Comment by soggycity** [comment]
Now Chrome won't let the extension run at all. Front end development just got a lot harder.

![Image](https://github.com/user-attachments/assets/f50bc85f-45e3-4dff-9674-dcc63dbe9e74)

**Comment by neilcsmith-net** [comment]
The extension has never been released by the Apache project.  We should probably add to the list of legacy things that no-one is working on and need removing?

**Comment by mhalachev** [comment]
If it has been turned off by Chrome, the extension can be enabled back manually in `chrome://extensions/`.

It actually still works, although with some workarounds. It would be great to bring it back as previously discussed.

> @mbien Would it be appropriate to get in touch with Jan Stola regarding the extension refactoring?
> @mhalachev If you want to ask to please refactor the plugin then probably not since too much time passed already ;)

No, not to refactor but at least to give us some direction :)

> @neilcsmith-net @ebarboni do we have access to the account which manages the plugin at the chrome store? If not: we should at some point try to take the plugin down since its not ideal when it points back to NB and we can't upload a new release.

If there is no access, perhaps release it as a new one?

**Comment by jepsar** [comment]
Is there any workaround on how to install it?

**Comment by mhalachev** [comment]
@jepsar It can be [loaded unpacked](https://developer.chrome.com/docs/extensions/get-started/tutorial/hello-world#load-unpacked) in Chrome extensions developer mode.

The extension itself is in  [webcommon/extbrowser.chrome/plugins/chrome
](https://github.com/apache/netbeans/tree/master/webcommon/extbrowser.chrome/plugins/chrome)

**Comment by toel--** [comment]
Following the instruction from @mhalachev, I could get the plugin working with reload on save at least:
https://lokato.se/temp/netbeans_chrome_plugin.zip
or if you just want to browse the files:
https://lokato.se/temp/netbeans_chrome_plugin/

**Comment by mpizeli** [comment]
@toel-- hi, please how did you manage to make it work ?
I have loaded it unpacked, but i get errors this.socket = new WebSocket(this.serverURL());

**Comment by lucadepascale** [comment]
due to some changes in chrome [policies](https://support.google.com/chrome_webstore/answer/2664769#unsupported_extensions) the connector has been blocked in chrome and cannot be nor found by search neither installed from the webstore. help, i'm lost without it!! and i'm sure i'm not the only one! 😆

**Comment by mpizeli** [comment]
since v3 never worked for me, i use chromium 1259956 (https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Win_x64/1259956/) and i manually install the crx from https://www.crx4chrome.com/crx/1439/

**Comment by mhalachev** [comment]
Recent Chrome versions no longer load extensions using Manifest V2. The "duct-tape" workaround mentioned above can still be side-loaded — at least, it allows inspecting the DOM of a simple web page in NetBeans, but the extension’s UI remains non-functional.

**Comment by pmatabiau** [comment]
Hi,
I've been waiting a long time for an update to NetBeans Connector, but it seems it will never come.
As a workaround, there's the [XDebug extension](https://chromewebstore.google.com/detail/oiofkammbajfehgpleginfomeppgnglk?utm_source=item-share-cb), which is a great replacement. You just need to define "netbeans" in its options (IDE key).
So far, everything seems to work (especially stepping through PHP), even better, as it seems more stable.
